internal/provider/resource_managed_identities.go

```diff
@@ -0,0 +1,433 @@
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state ManagedIdentityModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the managed identity from Tharsis.
+	found, err := t.client.ManagedIdentity.GetManagedIdentity(ctx, &ttypes.GetManagedIdentityInput{
+		ID: state.ID.ValueString(),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error reading managed identity",
+			err.Error(),
+		)
+		return
+	}
+
+	// Copy the from-Tharsis struct to the state.
+	copyManagedIdentity(*found, &state)
+
+	// Set the refreshed state.
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (t *managedIdentityResource) Update(ctx context.Context,
+	req resource.UpdateRequest, resp *resource.UpdateResponse) {
+
+	// Get the current state for its ID.
+	var state ManagedIdentityModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Retrieve values from plan for the description and data.
+	var plan ManagedIdentityModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	encodedData, err := encodeDataString(plan.Type,
+		managedIdentityDataInput{
+			AWSRole:       plan.AWSRole.ValueString(),
+			AzureClientID: plan.AzureClientID.ValueString(),
+			AzureTenantID: plan.AzureTenantID.ValueString(),
+		})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error encoding managed identity data field",
+			err.Error(),
+		)
+		return
+	}
+
+	// Update the managed identity via Tharsis.
+	// The ID is used to find the record to update.
+	// The description and data are modified.
+	updated, err := t.client.ManagedIdentity.UpdateManagedIdentity(ctx,
+		&ttypes.UpdateManagedIdentityInput{
+			ID:          state.ID.ValueString(),
+			Description: plan.Description.ValueString(),
+			Data:        encodedData,
+		})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error updating managed identity",
+			err.Error(),
+		)
+		return
+	}
+
+	// Copy all fields returned by Tharsis back into the plan.
+	copyManagedIdentity(*updated, &plan)
+
+	// Set the response state to the fully-populated plan.
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (t *managedIdentityResource) Delete(ctx context.Context,
+	req resource.DeleteRequest, resp *resource.DeleteResponse) {
+
+	// Get the current state.
+	var state ManagedIdentityModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Delete the managed identity via Tharsis.
+	// The ID is used to find the record to delete.
+	err := t.client.ManagedIdentity.DeleteManagedIdentity(ctx,
+		&ttypes.DeleteManagedIdentityInput{
+			ID: state.ID.ValueString(),
+		})
+	if err != nil {
```

Comment:
This should ignore errors that indicate the resource no longer exists
https://developer.hashicorp.com/terraform/plugin/framework/resources/delete#recommendations